### PR TITLE
[DynamoDB] measure, publish and alarm on high resolution consumed capacity

### LINKED ISSUE #153: [DynamoDB] Optionally request consumed capacity with all DynamoDB requests

### PR DESCRIPTION
**Pull request by amcp**
Fixes #153, #154 and #157

### CODE REVIEW

File: ndbench-dynamodb-plugins/build.gradle [hunk 1]
```diff
@@ -8,6 +8,8 @@ repositories {
 dependencies {
     compile project(':ndbench-api')
     compile "com.amazonaws:aws-java-sdk-dynamodb:latest.release"
+    compile "com.amazonaws:aws-java-sdk-cloudwatch:latest.release"
```
Comment (amcp) [inline ndbench-dynamodb-plugins/build.gradle]: are both necessary? or only one?

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 2]
```diff
@@ -122,34 +152,63 @@ public void init(DataGenerator dataGenerator) throws Exception {
...
+
         if (!Strings.isNullOrEmpty(this.config.getEndpoint())) {
             Preconditions.checkState(!Strings.isNullOrEmpty(config.getRegion()),
                     "If you set the endpoint you must set the region");
-            builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(config.getEndpoint(), config.getRegion()));
+            // dynamodb region is overriden. for high resolution consumption metrics to appear next to standard
+            // resolution service metrics, cloudwatch metrics need to be sent to the same region
+            cloudWatchClientBuilder.withRegion(Regions.fromName(config.getRegion()));
+            dynamoDbBuilder.withEndpointConfiguration(
+                    new AwsClientBuilder.EndpointConfiguration(config.getEndpoint(),config.getRegion()));
```
Comment (amcp) [inline ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: minor, add space after comma

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 3]
```diff
@@ -201,8 +276,11 @@ public String writeSingle(String key) throws Exception {
             final Item item = new Item()
                     .withPrimaryKey(partitionKeyName, key)
                     .withString(ATTRIBUTE_NAME, this.dataGenerator.getRandomValue());
+            final PutItemSpec spec = new PutItemSpec()
+                    .withItem(item)
+                    .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL);
             // Write the item to the table
-            final PutItemOutcome outcome = table.putItem(item);
+            final PutItemOutcome outcome = table.putItem(spec);
```
Comment (amcp) [inline ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: rebase on vanilla SDK PR when merged #156

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 4]
```diff
@@ -401,4 +541,35 @@ private void deleteTable() {
...
+            timer = Executors.newFixedThreadPool(1);
+            timer.submit((Callable<Void>) () -> {
+                while (!Thread.currentThread().isInterrupted()) {
+                    final Date now = Date.from(Instant.now());
+                    //publish to cloudwatch every second
+                    cloudWatch.putMetricData(new PutMetricDataRequest().withMetricData(
+                            createCuMetricDatumAndResetCounter(now, readConsumed, ND_BENCH_DYNAMO_DB_CONSUMED_RCU),
+                            createCuMetricDatumAndResetCounter(now, writeConsumed, ND_BENCH_DYNAMO_DB_CONSUMED_WCU)
+                    ));
+                    Thread.sleep(1000); //TODO make reporting interval pollablesdf
```
Comment (amcp) [inline ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: s/pollablesdf/configurable/

File: ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java [hunk 5]
```diff
@@ -401,4 +541,35 @@ private void deleteTable() {
...
+                }
+                return null;
+            });
+            timerRef.set(timer);
+        }
+    }
+
+    private MetricDatum createCuMetricDatumAndResetCounter(Date now, AtomicDouble count, String name) {
```
Comment (amcp) [inline ndbench-dynamodb-plugins/src/main/java/com/netflix/ndbench/plugin/dynamodb/DynamoDBKeyValue.java]: rename to createCapacityUnitMetricDatumAndResetCounter

**Comment by amcp** [comment]
rebase on CLI branch

**Comment by amcp** [comment]
I rebased on the cli branch for easier testing.

**Comment by amcp** [comment]
Setup:
```bash
sudo yum update -y \
&& sudo yum upgrade -y \
&& sudo yum install -y java-1.8.0-openjdk-devel git emacs \
&& sudo yum remove -y java-1.7.0-openjdk \
&& git clone https://github.com/amcp/ndbench.git \
&& cd ndbench \
&& git checkout consumed \
&& ./gradlew build
```

Test 1:
```bash
DISCOVERY_ENV=AWS ./gradlew run \
-Dndbench.config.cli.clientName=DynamoDBKeyValue \
-Dndbench.config.cli.bulkSize=1 \
-Dndbench.config.cli.timeoutMillis=360000 \
-Dndbench.config.dataSize=1000 \
-Dndbench.config.numKeys=1000000 \
-Dndbench.config.readEnabled=true \
-Dndbench.config.readRateLimit=25 \
-Dndbench.config.writeEnabled=true \
-Dndbench.config.writeRateLimit=25 \
-Dndbench.config.dynamodb.consistentread=true \
-Dndbench.config.dynamodb.maxRetries=0 \
-Dndbench.config.dynamodb.requestTimeout=100 \
-Dndbench.config.dynamodb.maxConnections=1 \
-Dndbench.config.dynamodb.numReaders=1 \
-Dndbench.config.dynamodb.numWriters=1 \
-Dndbench.config.dynamodb.programmableTables=true \
-Dndbench.config.dynamodb.publishHighResolutionConsumptionMetrics=true \
&> output
```

Test 2:
```bash
DISCOVERY_ENV=AWS ./gradlew run \
-Dndbench.config.cli.clientName=DynamoDBKeyValue \
-Dndbench.config.cli.bulkSize=1 \
-Dndbench.config.cli.timeoutMillis=360000 \
-Dndbench.config.dataSize=1000 \
-Dndbench.config.numKeys=1000000 \
-Dndbench.config.readEnabled=true \
-Dndbench.config.readRateLimit=25 \
-Dndbench.config.writeEnabled=true \
-Dndbench.config.writeRateLimit=25 \
-Dndbench.config.dynamodb.consistentread=true \
-Dndbench.config.dynamodb.maxRetries=0 \
-Dndbench.config.dynamodb.requestTimeout=100 \
-Dndbench.config.dynamodb.maxConnections=1 \
-Dndbench.config.dynamodb.numReaders=1 \
-Dndbench.config.dynamodb.numWriters=1 \
-Dndbench.config.dynamodb.programmableTables=true \
-Dndbench.config.dynamodb.publishHighResolutionConsumptionMetrics=true \
-Dndbench.config.dynamodb.alarmOnHighResolutionConsumptionMetrics=true \
&> output
```

**Comment by amcp** [comment]
screen shot 2018-04-16 at 9 54 17 am
test 1 seems to work, moving on to the alarm.

**Comment by amcp** [comment]
Verified the alarm works:
screen shot 2018-04-16 at 12 54 48 pm

**Comment by amcp** [comment]
I confirmed that the high-resolution alarm went into ALARM state about 1 minute after the load started. ALARM is on 10-second readings of SUM(Capacity units) being above 40 (10 second average of 4 TPS) for 5 consecutive data points (so takes about a minute to fire the alarm with a load).

File: ndbench-dynamodb-plugins/build.gradle (hunk 1 above)
Comment (ipapapa) [reply ndbench-dynamodb-plugins/build.gradle]: I do not believe we need the cloudwatch. I do not recall adding it.

**Comment by ipapapa** [comment]
There are a few conflicting files. I also need to understand a little better what is the high-resolution alarm. Can you please provide a brief overview?

**Comment by amcp** [comment]
A high-resolution alarm is one that is configured to fire on threshold breaches of high-resolution metrics. See this [announcement](https://aws.amazon.com/about-aws/whats-new/2017/07/amazon-cloudwatch-introduces-high-resolution-custom-metrics-and-alarms/). DynamoDB only publishes 1 minute consumed capacity metrics. By publishing high resolution consumed capacity metrics on the client side, you can react and alarm on spikes in load much quicker. Note, these client side consumption metrics are not compatible with application autoscaling, so you would have to write your own lambda  function to handle the breach of the high resolution alarm by up-provisioning the table.

**Comment by amcp** [comment]
@ipapapa I added some comments to document the purpose of this change, and an [issue to investigate whether the high resolution alarm is compatible with application autoscaling](https://github.com/Netflix/ndbench/issues/190).

**Comment by ipapapa** [comment]
I like the fact that you added the comments in the PR. It would be useful to move some of these comments into the `README.md` as well of this plugin.

**Pull request merged**